src/components/templates/Wrapper.tsx

```diff
@@ -0,0 +1,398 @@
+import {
+  DarkMode,
+  Flex,
+  LinkBox,
+  SimpleGrid,
+  Image,
+  Box,
+  LinkOverlay,
+  InputGroup,
+  InputLeftElement,
+  Icon,
+  Input,
+  InputRightElement,
+  IconButton,
+} from "@chakra-ui/react";
+import { HomepageFooter } from "components/footer/Footer";
+import { GetStartedSection } from "components/homepage/sections/GetStartedSection";
+import { NewsletterSection } from "components/homepage/sections/NewsletterSection";
+import { HomepageTopNav } from "components/product-pages/common/Topnav";
+import { HomepageSection } from "components/product-pages/homepage/HomepageSection";
+import { NextSeo } from "next-seo";
+import { Heading, Text, TrackedLink } from "tw-components";
+import { ReactNode, useRef, useState } from "react";
+import FilterMenu from "./FilterMenu";
+import { FiSearch, FiX } from "react-icons/fi";
+import { useRouter } from "next/router";
+import { TEMPLATE_DATA, TemplateCardProps } from "data/templates/templates";
+import { TEMPLATE_TAGS, TemplateTagId } from "data/templates/tags";
+
+type TemplateWrapperProps = {
+  title: string;
+  description: string;
+  data: TemplateCardProps[];
+  children: ReactNode;
+  showFilterMenu?: boolean;
+  _defaultTagIds?: TemplateTagId[];
+  _defaultKeyword?: string;
+};
+
+export function filterTemplates(
+  tagIds: TemplateTagId[],
+  keyword: string,
+  showFilterMenu?: boolean,
+  defaultData?: TemplateCardProps[],
+) {
+  let _templates = defaultData || TEMPLATE_DATA;
+  if (tagIds.length && showFilterMenu) {
+    _templates = TEMPLATE_DATA.filter((item) =>
+      tagIds.every((tagId) => item.tags.includes(tagId)),
+    );
+  }
+
+  // Don't search if there's only one letter
+  if (keyword && keyword.length >= 2) {
+    const _keyword = keyword.toLowerCase();
+    _templates = _templates.filter(
+      (template) =>
+        template.tags.includes(_keyword as TemplateTagId) ||
+        template.keywords?.includes(_keyword as TemplateTagId) ||
+        template.title.toLowerCase().includes(_keyword) ||
+        template.description.toLowerCase().includes(_keyword),
+    );
+  }
+  return _templates;
+}
+
+export default function TemplateWrapper(props: TemplateWrapperProps) {
+  const {
+    title,
+    description,
+    data,
+    children,
+    showFilterMenu,
+    _defaultTagIds,
+    _defaultKeyword,
+  } = props;
+  const router = useRouter();
+  const inputRef = useRef<HTMLInputElement>(null);
+  const [selectedTags, setSelectedTags] = useState<TemplateTagId[]>(
+    _defaultTagIds || [],
+  );
+  const [keyword, setKeyword] = useState<string>(_defaultKeyword || "");
```

Comment:
if ur updating the url you should also read from the url to initialize the state